cmd/tink-worker/internal/worker.go

```diff
@@ -45,28 +48,162 @@ type WorkflowMetadata struct {
-	workerID := os.Getenv("WORKER_ID")
-	if workerID == "" {
-		return errors.New("required WORKER_ID")
+type Worker struct {
+	client         pb.WorkflowSvcClient
+	regConn        *RegistryConnDetails
+	registryClient *client.Client
+	logger         log.Logger
+	registry       string
+	retries        int
+	retryInterval  time.Duration
+	maxSize        int64
+}
+
+func NewWorker(client pb.WorkflowSvcClient, regConn *RegistryConnDetails, logger log.Logger, registry string, retries int, retryInterval time.Duration, maxSize int64) *Worker {
+	registryClient, err := regConn.NewClient()
+	if err != nil {
+		panic(err)
+	}
+	return &Worker{
+		client:         client,
+		regConn:        regConn,
+		registryClient: registryClient,
+		logger:         logger,
+		registry:       registry,
+		retries:        retries,
+		retryInterval:  retryInterval,
+		maxSize:        maxSize,
 	}
+}
 
-	ctx := context.Background()
-	var err error
-	cli, err = initializeDockerClient()
+func (w *Worker) captureLogs(ctx context.Context, id string) {
+	reader, err := w.registryClient.ContainerLogs(context.Background(), id, types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     true,
+		Timestamps: false,
+	})
+	if err != nil {
+		panic(err)
+	}
+	defer reader.Close()
+
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		fmt.Println(scanner.Text())
+	}
+}
+
+func (w *Worker) execute(ctx context.Context, wfID string, action *pb.WorkflowAction) (pb.ActionState, error) {
+	l := w.logger.With("workflowID", wfID, "workerID", action.GetWorkerId(), "actionName", action.GetName(), "actionImage", action.GetImage())
+
+	cli := w.registryClient
+	err := w.regConn.pullImage(ctx, cli, action.GetImage())
+	if err != nil {
+		return pb.ActionState_ACTION_IN_PROGRESS, errors.Wrap(err, "DOCKER PULL")
+	}
+	id, err := w.createContainer(ctx, action.Command, wfID, action)
+	if err != nil {
+		return pb.ActionState_ACTION_IN_PROGRESS, errors.Wrap(err, "DOCKER CREATE")
+	}
+	l.With("containerID", id, "command", action.GetOnTimeout()).Info("container created")
+
+	var timeCtx context.Context
+	var cancel context.CancelFunc
+	if action.Timeout > 0 {
+		timeCtx, cancel = context.WithTimeout(ctx, time.Duration(action.Timeout)*time.Second)
+	} else {
+		timeCtx, cancel = context.WithTimeout(ctx, 1*time.Hour)
+	}
+	defer cancel()
+
+	err = startContainer(timeCtx, l, cli, id)
 	if err != nil {
-		return err
+		return pb.ActionState_ACTION_IN_PROGRESS, errors.Wrap(err, "DOCKER RUN")
+	}
+
+	failedActionStatus := make(chan pb.ActionState)
+
+	//capturing logs of action container in a go-routine
+	go w.captureLogs(ctx, id)
+
+	status, err := waitContainer(timeCtx, cli, id)
+	if err != nil {
+		rerr := removeContainer(ctx, l, cli, id)
+		if rerr != nil {
+			rerr = errors.Wrap(rerr, errRemoveContainer)
+			l.With("containerID", id).Error(rerr)
+			return status, rerr
+		}
+		return status, errors.Wrap(err, "DOCKER_WAIT")
 	}
+	rerr := removeContainer(ctx, l, cli, id)
```

Comment:
That makes sense.

I'm getting worried about the level of changes that I am making from the original (that this may become too complex to review): https://github.com/tinkerbell/tink/pull/287/files#diff-f776dc055922cd9d519a49a712a9d22fL75

Reply:
Would it be possible to break the changes down into more easily reviewable chunks?

Reply:
It looks like we are burying the original "wait" error on wait failure (without logging it), by replacing it with the removal error when removal fails.

If I don't defer the removal, I could wrap the errors and log both failures.  

I'll give the deer a shot.  Even though we may be buring the removal error, it will be deferred and logged, and the original "wait" error will be surfaced.  This should make the code and error easier to understand.

Reply:
The `removeContainer` calls at the bottom of the function don't even make it into the returned error (they are only logged).  I think defer is the right call based on the existing usage.

Reply:
I split up the commits a little, @detiber. 

The first commit now captures the `git mv` to `internal/`, which means that the second commit reflects the `actions.go` changes more succinctly.  I could do something similar to keep the "execute" function defined in the same original position as "executeAction" (likewise for capture logs) if that would help.

The third commit (fixup) represents the changes from your latest round of review and includes linter fixes needed to make `make verify` happy.  

The last commit moves all of the arguments from Persistent to local. I didn't venture down the other ways of getting viper to handle things. I think that can wait for a further cleanup, perhaps in your tink-server viper PR.